### ISSUE TITLE
6.1.3 Regression with selectors

### ISSUE THREAD

**Issue by kylecannon** [getsaf/shallow-render#60]
After upgrading to the latest shallow render, my tests started failing. Before, I was at 6.0.6.

Here is the error: 

```	Error: Template parse errors:
	Unexpected closing tag "__MockOfEmptyOutletComponent-selector". It may happen when the tag has already been closed by another tag. For more info see https://www.w3.org/TR/html5/syntax.html#closing-elements-that-have-implied-end-tags ("<__MockOfEmptyOutletComponent-selector>[ERROR ->]</__MockOfEmptyOutletComponent-selector>"): @0:39
	    at syntaxError (./node_modules/@angular/compiler/fesm5/compiler.js?:1275:17)
	    at TemplateParser.parse (./node_modules/@angular/compiler/fesm5/compiler.js?:15084:19)
	    at JitCompiler._parseTemplate (./node_modules/@angular/compiler/fesm5/compiler.js?:24272:37)
	    at JitCompiler._compileTemplate (./node_modules/@angular/compiler/fesm5/compiler.js?:24259:23)
	    at eval (./node_modules/@angular/compiler/fesm5/compiler.js?:24202:62)
	    at Set.forEach (<anonymous>)
	    at JitCompiler._compileComponents (./node_modules/@angular/compiler/fesm5/compiler.js?:24202:19)
	    at eval (./node_modules/@angular/compiler/fesm5/compiler.js?:24120:19)
	    at Object.then (./node_modules/@angular/compiler/fesm5/compiler.js?:1266:77)
```

Downgrading to 6.1.2 solves the issue.

Commit that seemed to cause it: https://github.com/getsaf/shallow-render/commit/07ac3a3553ae174b27c82caf4dd38e43e91f6543

**Comment by getsaf** [comment]
It looks like Shallow is attempting to mock the RouterModule internals. Angular provides a substitution module, are you using the `RouterTestingModule` in your tests? The routing stuff is a little weird because it uses a `forRoot` pattern + you need the testing module.

```typescript
// You probably want to export this ref from your module
// so we can give Shallow specs access to it too.
export const routerModuleRef = RouterModule.forRoot(routes);
@NgModule({
  imports: [routerModuleRef],
  /* ... */
})
class AppModule {}
```

Then, in your test:
```typescript
import { RouterTestingModule } from '@angular/router/testing';
import { routerModuleRef } from 'path/to/app.module.ts';

// ...
const testRoutes: Routes = [
  {path: 'home', component: class MyComponent {}}
];
new Shallow(MyComponent, MyModule)
  .replaceModule(routerModuleRef, RouterTestingModule.withRoutes(testRoutes)
  .dontMock(RouterTestingModule);
```

**Comment by kylecannon** [comment]
@getsaf 

Finally got around to trying you solution. I am unable to get this to work with your provided solution if you're using routing modules. Check here: https://stackblitz.com/edit/github-frqbzt-dek4h5?file=examples/component-with-routing.spec.ts

**Comment by kylecannon** [comment]
@getsaf please look at this when you get a chance. I don't know if you just missed this or not.

**Comment by getsaf** [comment]
Looked at that stackblitz.. I got it running by commenting out line 37. I'm not sure why you export `RoutingModule` on line 37 but if it's necessary, you will need to include it in your `dontMocks` so it doesn't get mocked out by Shallow.

If you can do without exporting it, I'd eliminate the `exports: [RoutingModule]`.

Let me know if this works for you or if there's something else going on that I'm not noticing.

**Comment by PaulDLW** [comment]
Would echo @kylecannon sentiment. Absolutely love using shallow, but it is a pain to export the routingModuleRef in lots of different places and update multiple tests to use it.

**Comment by getsaf** [comment]
@kylecannon Looks like it worked by simply replacing the `GoHomeRoutingModule` with the `RouterTestingModule` too, this way you won't have to export the `routerModuleRef` anymore.

```typescript
      .replaceModule(GoHomeRoutingModule, RouterTestingModule.withRoutes(routes));
```

https://stackblitz.com/edit/github-frqbzt-dek4h5?file=examples%2Fcomponent-with-routing.spec.ts

@PaulDLW Have you tried using `alwaysMock` or `alwaysProvide` to do the replacement globally? It seemed like always providing the [`RouterTestingModule`](https://angular.io/api/router/testing/RouterTestingModule) was a good solution.

**Comment by getsaf** [comment]
@villelahdenvuo, @kylecannon I dug into what happens when the `RouterModule` was being created and noticed that when using `RouterModule.forRoot()` it simply creates a `ModuleWithProviders` where the `ngModule` is `RouterModule`. The matching on `shallow.replaceModule` was very strict and wouldn't match the `ModuleWithReplacements` to the `RouterModule` but I can't think of any good reason NOT to treat this as a match.

I just made a PR for this fix (#77) and I will release a minor 7.1.0 since it could potentially break some folks under certain strange circumstances.
